charmcraft/commands/store/registry.py

```diff
@@ -203,6 +208,63 @@ def upload_manifest(self, manifest_data, reference):
         assert_response_ok(response, expected_status=201)
         logger.debug("Manifest uploaded OK")
 
+    def upload_blob(self, filepath, size, digest):
+        """Upload the blob from a file."""
+        # get the first URL to start pushing the blob
+        logger.debug("Getting URL to push the blob")
+        url = self._get_url("blobs/uploads/")
+        response = self._hit("POST", url)
+        assert_response_ok(response, expected_status=202)
+        upload_url = response.headers["Location"]
+        range_from, range_to_inclusive = [
+            int(x) for x in response.headers["Range"].split("-")
+        ]
+        logger.debug(
+            "Got upload URL ok with range %s-%s", range_from, range_to_inclusive
+        )
+        if range_from != 0:
+            raise CommandError("Server error: bad range received")
+        if range_to_inclusive == 0:
+            range_to_inclusive = -1
+
+        # start the chunked upload
+        from_position = range_to_inclusive + 1
+        with open(filepath, "rb") as fh:
+            fh.seek(from_position)
+            while True:
+                chunk = fh.read(CHUNK_SIZE)
+                if not chunk:
+                    break
+
+                end_position = from_position + len(chunk)
+                headers = {
+                    "Content-Length": str(len(chunk)),
+                    "Content-Range": "{}-{}".format(from_position, end_position),
+                    "Content-Type": OCTET_STREAM_MIMETYPE,
+                }
+                progress = 100 * end_position / size
+                print("Uploading.. {:.2f}%\r".format(progress), end="", flush=True)
```

Comment:
this makes me a bit uncomfortable if we intend to have an eventual common library (the printing from a package/library may not be as nice) and taking that into account or to start thinking about this would be nice

Reply:
This will be replaced when we have the library, of course.

I added a comment here in he code, and opened [an issue about it](https://github.com/canonical/charmcraft/issues/381).